scraper/parliament: Fix an issue with file:/// protocol URLs Fix #261

jdu opened the pull request:
# Description

This pull is to fix #261 see that issue for the specific details of the issue.

* Adds new `_is_valid_pdf` and `_is_valid_pdf_url` methods to `base_spider.py` which filter on headers as well as the URL fragment for selecting appropriate PDF files for download.

## Type of change

Please delete options that are not relevant.

- [x] :bug: Bug fix

# How Has This Been Tested?

`make docker-test`

# Checklist:

- [x] My code follows the style guidelines of this project (pep8 AND pyflakes)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes
- [x] If my PR aims to fix an issue, I referenced it using `#(issue)`

jdu commented:
This is WIP to get comments on implementation, there are a couple notes in addition:

1. We might want to update the tests to reflect catching the type fo scenario in #261 
2. This is explicitly requires that a download has an extension of 'pdf', do we know of any instances where a PDF is served from a route without and extension and the filename is passed as a header for instance (i.e. a route like `/files/documents/<uuid>` which has a header like `Content-Disposition: filename="some_document_2019.pdf"`

There's also an issue with filtering raised in #286 that came out of this.

SamDepardieu commented on reach/scraper/wsf_scraping/spiders/base_spider.py:
```diff
@@ -56,6 +60,61 @@ def _check_headers(self, response_headers,
+            extension = 'pdf'
+
+        if mimetype is None:
+            mimetype = b'application/pdf'
```
Why not settings these in the header?
```
def _is_valid_pdf(self, response, extension='pdf', mimetype=b'application/pdf'):
...
```

jdu replied:
Habit mostly, python default arguments are mutable as they're instantiated when defined and not at the point the actual method is called, so I've formed a habit of not specifying default arguments other than None to avoid polluting them down the road. 

In case you haven't seen this before and for anyone else reading this who might not have:

```
class SomeClass():
    def some_method(self, new_val, param=[]):
        param.append(new_val)
        print(param)

new_inst = SomeClass()
new_inst.some_method("HELLO")
new_inst.some_method("GOODBYE")
```

It's easy to assume that the above would output:

```
['HELLO']
['GOODBYE']
```

when it will actually output
```
['HELLO']
['HELLO', 'GOODBYE']
```

because `param` is mutable.

Instead of reasoning about every case I just built a habit out of always specifying None for keyword arguments and dealing with them explicitly in the methods body.

Can change in this case as it's not as much of an issue here, but really it just boiled down to habit...

hblanks replied:
:+1:  yes, finding out about mutable keyword args when debugging is no fun. I'm more ok with immutable keyword args not being None but agree, you'll save yourself some pain if you use None most of the time.

SamDepardieu replied:
That's good to know 👍

jdu commented:
Already done propagating changes to other scrapers, as well as some changes as the NICE site doesn't provide URLs with extensions for the PDF downloads so had to re-adjust the approach. Also updated some tests in line with these changes.

Just waiting on policytool-test to run the scrapers to make sure everything is working correctly then will update this PR.

jdu commented:
Refactored this a bit as the NICE scraper turned out to have URLs which don't contain a file extension for pdf. Updated some tests which didn't pass after these changes as the tests didn't specify a response with a pdf extension URL which caused them to fail the `is_vaild_pdf` and `is_valid_pdf_url` checks.

Have ran policy-test and scrapers run fine, as well as all tests passing in `make docker-test`

jdu commented:
Just to add to this, the parliament site emailed me back and have fixed at least the link in question which triggered the error.
